multiple definition link error when using multiple codegen units

I wanted to use @dotdash's LLVM assertions patch but now my stage1 libstd fails to link. Probably unrelated but here it is:

 ```go
rustc: arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/libstd
error: linking with `cc` failed: exit code: 1
note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-L" "/rust-master/arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib" "arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.0.o" "arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.1.o" "arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.2.o" "arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.3.o" "-o" "arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/libstd-9026086f.so" "arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.metadata.o" "-Wl,-O1" "-nodefaultlibs" "-L" "arm-unknown-linux-gnueabihf/rt" "-L" "/rust-master/arm-unknown-linux-gnueabihf/llvm/Release+Asserts/lib" "-L" "/rust-master/arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib" "-Wl,-Bstatic" "-Wl,--whole-archive" "-l" "backtrace" "-Wl,--no-whole-archive" "-Wl,-Bdynamic" "-l" "dl" "-l" "pthread" "-l" "gcc_s" "-Wl,--whole-archive" "/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FHfUQVlQsuoG/liballoc-9026086f.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FHfUQVlQsuoG/librustc_unicode-9026086f.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FHfUQVlQsuoG/liballoc_system-9026086f.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FHfUQVlQsuoG/liblibc-9026086f.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FHfUQVlQsuoG/librand-9026086f.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FHfUQVlQsuoG/libcore-9026086f.rlib" "-Wl,--no-whole-archive" "-l" "c" "-l" "m" "-l" "rt" "-shared" "-Wl,-rpath,/home/odroid/rust-nightly-sysalloc-assert/lib/rustlib/arm-unknown-linux-gnueabihf/lib" "-Wl,--enable-new-dtags" "-s" "-l" "compiler-rt"
````
```shell
note: /tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `_$LT$core..option..Option$LT$T$GT$$GT$::map::h58518f67af9bf04e':
collections.0.rs:(.text._ZN38_$LT$core..option..Option$LT$T$GT$$GT$3map17h58518f67af9bf04eE+0x0): multiple definition of `_$LT$core..option..Option$LT$T$GT$$GT$::map::h58518f67af9bf04e'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.0.o:std.0.rs:(.text._ZN38_$LT$core..option..Option$LT$T$GT$$GT$3map17h58518f67af9bf04eE+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$::new::h0c1cdac7fb2b416c':
collections.0.rs:(.text._ZN40_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$3new17h0c1cdac7fb2b416cE+0x0): multiple definition of `_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$::new::h0c1cdac7fb2b416c'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.3.o:std.3.rs:(.text._ZN40_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$3new17h0c1cdac7fb2b416cE+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$::ptr::h4e14ee1392d1411e':
collections.0.rs:(.text._ZN40_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$3ptr17h4e14ee1392d1411eE+0x0): multiple definition of `_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$::ptr::h4e14ee1392d1411e'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.0.o:std.0.rs:(.text._ZN40_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$3ptr17h4e14ee1392d1411eE+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$::double::heaf5b97a8495eb9e':
collections.0.rs:(.text._ZN40_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$6double17heaf5b97a8495eb9eE+0x0): multiple definition of `_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$::double::heaf5b97a8495eb9e'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.3.o:std.3.rs:(.text._ZN40_$LT$alloc..raw_vec..RawVec$LT$T$GT$$GT$6double17heaf5b97a8495eb9eE+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `core::mem::uninitialized::he8ba150ad18eedcf':
collections.0.rs:(.text._ZN4core3mem13uninitialized17he8ba150ad18eedcfE+0x0): multiple definition of `core::mem::uninitialized::he8ba150ad18eedcf'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.0.o:std.0.rs:(.text._ZN4core3mem13uninitialized17he8ba150ad18eedcfE+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `core::mem::uninitialized::hfa1a47f766d26f10':
collections.0.rs:(.text._ZN4core3mem13uninitialized17hfa1a47f766d26f10E+0x0): multiple definition of `core::mem::uninitialized::hfa1a47f766d26f10'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.3.o:std.3.rs:(.text._ZN4core3mem13uninitialized17hfa1a47f766d26f10E+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `core::mem::transmute_copy::h5f22bd4094d00101':
collections.0.rs:(.text._ZN4core3mem14transmute_copy17h5f22bd4094d00101E+0x0): multiple definition of `core::mem::transmute_copy::h5f22bd4094d00101'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.0.o:std.0.rs:(.text._ZN4core3mem14transmute_copy17h5f22bd4094d00101E+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `core::mem::transmute_copy::h75f84324b8ee3166':
collections.0.rs:(.text._ZN4core3mem14transmute_copy17h75f84324b8ee3166E+0x0): multiple definition of `core::mem::transmute_copy::h75f84324b8ee3166'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.3.o:std.3.rs:(.text._ZN4core3mem14transmute_copy17h75f84324b8ee3166E+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `core::mem::forget::h502109c553961f51':
collections.0.rs:(.text._ZN4core3mem6forget17h502109c553961f51E+0x0): multiple definition of `core::mem::forget::h502109c553961f51'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.0.o:std.0.rs:(.text._ZN4core3mem6forget17h502109c553961f51E+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `core::mem::size_of::hfb509be3c362d215':
collections.0.rs:(.text._ZN4core3mem7size_of17hfb509be3c362d215E+0x0): multiple definition of `core::mem::size_of::hfb509be3c362d215'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.0.o:std.0.rs:(.text._ZN4core3mem7size_of17hfb509be3c362d215E+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `core::mem::align_of::h2a452ce69533a17c':
collections.0.rs:(.text._ZN4core3mem8align_of17h2a452ce69533a17cE+0x0): multiple definition of `core::mem::align_of::h2a452ce69533a17c'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.0.o:std.0.rs:(.text._ZN4core3mem8align_of17h2a452ce69533a17cE+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `core::ptr::_$LT$impl$u20$$BP$mut$u20$T$GT$::is_null::h3fe8e4e4a8815d84':
collections.0.rs:(.text._ZN4core3ptr31_$LT$impl$u20$$BP$mut$u20$T$GT$7is_null17h3fe8e4e4a8815d84E+0x0): multiple definition of `core::ptr::_$LT$impl$u20$$BP$mut$u20$T$GT$::is_null::h3fe8e4e4a8815d84'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.0.o:std.0.rs:(.text._ZN4core3ptr31_$LT$impl$u20$$BP$mut$u20$T$GT$7is_null17h3fe8e4e4a8815d84E+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `core::ptr::_$LT$impl$u20$$BP$const$u20$T$GT$::offset::he051aa5e5b8f8bfb':
collections.0.rs:(.text._ZN4core3ptr33_$LT$impl$u20$$BP$const$u20$T$GT$6offset17he051aa5e5b8f8bfbE+0x0): multiple definition of `core::ptr::_$LT$impl$u20$$BP$const$u20$T$GT$::offset::he051aa5e5b8f8bfb'
arm-unknown-linux-gnueabihf/stage1/lib/rustlib/arm-unknown-linux-gnueabihf/lib/std-9026086f.0.o:std.0.rs:(.text._ZN4core3ptr33_$LT$impl$u20$$BP$const$u20$T$GT$6offset17he051aa5e5b8f8bfbE+0x0): first defined here
/tmp/rustc.FHfUQVlQsuoG/libcollections-9026086f.rlib(collections-9026086f.0.o): In function `core::ptr::_$LT$impl$u20$$BP$const$u20$T$GT$::is_null::h630d425861f40502':
collections.0.rs:(.text._ZN4core3ptr33_$LT$impl$u20$$BP$const$u20$T$GT$7is_null17h630d425861f40502E+0x0): multiple definition of `core::ptr::_$LT$impl$u20$$BP$const$u20$T$GT$::is_null::h630d425861f40502'
````

and so on. Configured with `--enable-optimize --disable-jemalloc --enable-llvm-assertions` and `RUSTFLAGS='-C codegen-units=4 -C link-args=-s'`

Are all these errors about memory management  functions? (`--disable-jemalloc` broken again?)

--- a/src/librustc_llvm/lib.rs
+++ b/src/librustc_llvm/lib.rs
@@ -2125,6 +2125,9 @@ extern {
     pub fn LLVMRustFreeOperandBundleDef(Bundle: OperandBundleDefRef);
 
     pub fn LLVMRustPositionBuilderAtStart(B: BuilderRef, BB: BasicBlockRef);
+
+    pub fn LLVMRustSetComdat(M: ModuleRef, V: ValueRef, Name: *const c_char);
+    pub fn LLVMRustUnsetComdat(V: ValueRef);
 }
 
 // LLVM requires symbols from this library, but apparently they're not printed
@@ -2146,6 +2149,24 @@ pub fn SetFunctionCallConv(fn_: ValueRef, cc: CallConv) {
 pub fn SetLinkage(global: ValueRef, link: Linkage) {
     unsafe {
         LLVMSetLinkage(global, link as c_uint);
+    }
+}
+
+// Externally visible symbols that might appear in multiple translation units need to appear in
+// their own comdat section so that the duplicates can be discarded at link time. This can for
+// example happen for generics when using multiple codegen units. This function simply uses the
+// value's name as the comdat value to make sure that it is in a 1-to-1 relationship to the
+// function.
+// For more details on COMDAT sections see e.g. http://www.airs.com/blog/archives/52
+pub fn SetUniqueComdat(llmod: ModuleRef, val: ValueRef) {
+    unsafe {
+        LLVMRustSetComdat(llmod, val, LLVMGetValueName(val));
+    }
+}
+
+pub fn UnsetComdat(val: ValueRef) {
+    unsafe {
+        LLVMRustUnsetComdat(val);
     }
 }
 

--- a/src/librustc_trans/base.rs
+++ b/src/librustc_trans/base.rs
@@ -2215,18 +2215,27 @@ pub fn update_linkage(ccx: &CrateContext,
         }
     }
 
-    match id {
-        Some(id) if ccx.reachable().contains(&id) => {
+    let (is_reachable, is_generic) = if let Some(id) = id {
+        (ccx.reachable().contains(&id), false)
+    } else {
+        (false, true)
+    };
+
+    // We need external linkage for items reachable from other translation units, this include
+    // other codegen units in case of parallel compilations.
+    if is_reachable || ccx.sess().opts.cg.codegen_units > 1 {
+        if is_generic {
+            // This only happens with multiple codegen units, in which case we need to use weak_odr
+            // linkage because other crates might expose the same symbol. We cannot use
+            // linkonce_odr here because the symbol might then get dropped before the other codegen
+            // units get to link it.
+            llvm::SetUniqueComdat(ccx.llmod(), llval);
+            llvm::SetLinkage(llval, llvm::WeakODRLinkage);
+        } else {
             llvm::SetLinkage(llval, llvm::ExternalLinkage);
-        },
-        _ => {
-            // `id` does not refer to an item in `ccx.reachable`.
-            if ccx.sess().opts.cg.codegen_units > 1 {
-                llvm::SetLinkage(llval, llvm::ExternalLinkage);
-            } else {
-                llvm::SetLinkage(llval, llvm::InternalLinkage);
-            }
-        },
+        }
+    } else {
+        llvm::SetLinkage(llval, llvm::InternalLinkage);
     }
 }
 
@@ -2547,8 +2556,10 @@ fn internalize_symbols(cx: &SharedCrateContext, reachable: &HashSet<&str>) {
         // then give it internal linkage.
         for ccx in cx.iter() {
             for val in iter_globals(ccx.llmod()).chain(iter_functions(ccx.llmod())) {
+                let linkage = llvm::LLVMGetLinkage(val);
                 // We only care about external definitions.
-                if !(llvm::LLVMGetLinkage(val) == llvm::ExternalLinkage as c_uint &&
+                if !((linkage == llvm::ExternalLinkage as c_uint ||
+                      linkage == llvm::WeakODRLinkage as c_uint) &&
                      llvm::LLVMIsDeclaration(val) == 0) {
                     continue;
                 }
@@ -2560,6 +2571,7 @@ fn internalize_symbols(cx: &SharedCrateContext, reachable: &HashSet<&str>) {
                    !reachable.contains(str::from_utf8(&name).unwrap()) {
                     llvm::SetLinkage(val, llvm::InternalLinkage);
                     llvm::SetDLLStorageClass(val, llvm::DefaultStorageClass);
+                    llvm::UnsetComdat(val);
                 }
             }
         }

--- a/src/librustc_trans/monomorphize.rs
+++ b/src/librustc_trans/monomorphize.rs
@@ -123,7 +123,6 @@ pub fn monomorphic_fn<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
             ref attrs, node: hir::ImplItemKind::Method(
                 hir::MethodSig { ref decl, .. }, ref body), ..
         }) => {
-            base::update_linkage(ccx, lldecl, None, base::OriginalTranslation);
             attributes::from_fn_attrs(ccx, attrs, lldecl);
 
             let is_first = !ccx.available_monomorphizations().borrow()
@@ -133,12 +132,14 @@ pub fn monomorphic_fn<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
             }
 
             let trans_everywhere = attr::requests_inline(attrs);
-            if trans_everywhere && !is_first {
-                llvm::SetLinkage(lldecl, llvm::AvailableExternallyLinkage);
-            }
-
             if trans_everywhere || is_first {
+                let origin = if is_first { base::OriginalTranslation } else { base::InlinedCopy };
+                base::update_linkage(ccx, lldecl, None, origin);
                 trans_fn(ccx, decl, body, lldecl, psubsts, fn_node_id);
+            } else {
+                // We marked the value as using internal linkage earlier, but that is illegal for
+                // declarations, so switch back to external linkage.
+                llvm::SetLinkage(lldecl, llvm::ExternalLinkage);
             }
         }
 

--- a/src/rustllvm/RustWrapper.cpp
+++ b/src/rustllvm/RustWrapper.cpp
@@ -1189,3 +1189,16 @@ extern "C" void LLVMRustPositionBuilderAtStart(LLVMBuilderRef B, LLVMBasicBlockR
     auto point = unwrap(BB)->getFirstInsertionPt();
     unwrap(B)->SetInsertPoint(unwrap(BB), point);
 }
+
+extern "C" void LLVMRustSetComdat(LLVMModuleRef M, LLVMValueRef V, const char *Name) {
+    Triple TargetTriple(unwrap(M)->getTargetTriple());
+    GlobalObject *GV = unwrap<GlobalObject>(V);
+    if (!TargetTriple.isOSBinFormatMachO()) {
+        GV->setComdat(unwrap(M)->getOrInsertComdat(Name));
+    }
+}
+
+extern "C" void LLVMRustUnsetComdat(LLVMValueRef V) {
+    GlobalObject *GV = unwrap<GlobalObject>(V);
+    GV->setComdat(nullptr);
+}

--- a/src/test/auxiliary/cgu_test.rs
+++ b/src/test/auxiliary/cgu_test.rs
@@ -1,0 +1,16 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// no-prefer-dynamic
+// compile-flags: --crate-type=lib
+
+pub fn id<T>(t: T) -> T {
+  t
+}

--- a/src/test/auxiliary/cgu_test_a.rs
+++ b/src/test/auxiliary/cgu_test_a.rs
@@ -1,0 +1,25 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// no-prefer-dynamic
+// compile-flags: -Ccodegen-units=2 --crate-type=lib
+
+extern crate cgu_test;
+
+pub mod a {
+    pub fn a() {
+        ::cgu_test::id(0);
+    }
+}
+pub mod b {
+    pub fn a() {
+        ::cgu_test::id(0);
+    }
+}

--- a/src/test/auxiliary/cgu_test_b.rs
+++ b/src/test/auxiliary/cgu_test_b.rs
@@ -1,0 +1,25 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// no-prefer-dynamic
+// compile-flags: -Ccodegen-units=2 --crate-type=lib
+
+extern crate cgu_test;
+
+pub mod a {
+    pub fn a() {
+        ::cgu_test::id(0);
+    }
+}
+pub mod b {
+    pub fn a() {
+        ::cgu_test::id(0);
+    }
+}

--- a/src/test/run-pass/issue-32518.rs
+++ b/src/test/run-pass/issue-32518.rs
@@ -1,0 +1,22 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// no-prefer-dynamic
+// aux-build:cgu_test.rs
+// aux-build:cgu_test_a.rs
+// aux-build:cgu_test_b.rs
+
+extern crate cgu_test_a;
+extern crate cgu_test_b;
+
+fn main() {
+    cgu_test_a::a::a();
+    cgu_test_b::a::a();
+}

